platforms/avr/drivers/ws2812_spi_int.c

```diff
@@ -0,0 +1,96 @@
+ */
+
+/* This is an interrupt-driven SPI driver for the
+ * WS2812/SK6812 RGB leds, for the ATmega32U.
+ *
+ * It sets SPI to 8MHz and sends 11100000 for a zero
+ * and 11111110 for a one. The time used by the ISR
+ * to prepare the next bit ensures the timming after
+ * each bit.
+ *
+ * It requires a 16MHz CPU clock and the pins PB0 and PB1
+ * unused (they are part of the ISP interface, so it's not
+ * uncommon), and the leds connected to the PB2 pin. Also,
+ * it can manage a maximum of 85 leds.
+ *
+ * It uses 40 bytes more than the WS2812 driver, but being interrupt
+ * based, allows to multiplex the LEDs with other tasks.
+ */
+
+#include <avr/interrupt.h>
+#include "rgblight.h"
+
+uint8_t *led_data;
+uint8_t  nbytes;
+uint8_t  remaining_bits;
+
+void ws2812_setleds(LED_TYPE *ledarray, uint16_t leds) {
+    led_data       = (uint8_t *)ledarray;
+    nbytes         = 3 * leds;
+    remaining_bits = 8;
+    // Initialize SPI for the WS2812/SK6812 driver
+    // Set PB2 and PB0 as output (not doing it freezes the SPI)
+    DDRB |= (1 << PB2);
+    DDRB |= (1 << PB0);
+    // Enable SPI, Master, set clock rate fck/2, MODE1, and enable interrupts
+    SPSR = (1 << SPI2X);
+    SPCR = (1 << SPIE) | (1 << SPE) | (1 << MSTR) | (1 << CPHA);
+    SPDR = 0; // send a byte to trigger an interrupt and begin the process
+}
+
+ISR(SPI_STC_vect, ISR_NAKED) {                //                                    "naked" ISR; we have to preserve any register that we use.
+    asm volatile("        push r25\n\t"       //                                    Using assembly code is better to reduce the duration of the ISR to the bare minimum.
+                 "        in   r25, 0x3F\n\t" //                                    preserve the SREG status register
+                 "        push r25\n\t"
+                 "        push r24\n\t" //                                          preserve r24 and r25
+                 "        push r30\n\t"
+                 "        push r31\n\t" //                                          preserve Z (r30/r31)
+                 "        lds  r30, led_data\n\t"
+                 "        lds  r31, led_data+1\n\t"     // uint8_t *Z = led_data;   read the address of the byte to send
+                 "        ld   r24, Z\n\t"              // if (*Z & 0x80)           read the byte to send
+                 "        ldi  r25, 0xE0\n\t"           //     r25 = 0xE0;          11100000 -> zero bit (375ns up)
+                 "        sbrc r24, 7\n\t"              // else
+                 "        ldi  r25, 0xFE\n\t"           //     r25 = 0xFE;          11111110 -> one bit (875ns up)
+                 "        add  r24, r24\n\t"            // *Z <<= 2;                rotate the current byte one bit to the left
+                 "        st   Z+, r24\n\t"             // Z++;                     store it in memory and point to the next byte (the incremented value won't be
```

Comment:
So apparently this code modifies the passed `leddata` array; this would make it incompatible with RGB Matrix (the code there assumes that the array passed to `ws2812_setleds()` is not modified by the driver).

Reply:
Fixed.

Reply:
So apparently you replaced a plain logical shift (which just loses bits that were shifted out) with a rotate (which restores the original value after repeating the operation 8 times).  However, the issue is that on each interrupt the rotate is done only once, and the intermediate state remains visible to the code outside the interrupt handler; although the QMK code should not read that state, it may write to the LED buffer, and if that write touches a byte that was currently shifted out, that byte would be corrupted.

(Actually even if that part would be fixed, there would be another problem — the driver could send some partially updated state to the LEDs; but at least this case won't result in a persistently wrong state, unlike the case of overwriting a partially rotated byte.)